modules/process/manager_stacktraces.go

```diff
@@ -0,0 +1,256 @@
+// Copyright 2022 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package process
+
+import (
+	"fmt"
+	"io"
+	"runtime/pprof"
+	"sort"
+	"time"
+
+	"github.com/google/pprof/profile"
+)
+
+// StackEntry is an entry on a stacktrace
+type StackEntry struct {
+	Function string
+	File     string
+	Line     int
+}
+
+// Label represents a pprof label assigned to goroutine stack
+type Label struct {
+	Name  string
+	Value string
+}
+
+// Stack is a stacktrace relating to a goroutine. (Multiple goroutines may have the same stacktrace)
+type Stack struct {
+	Count       int64 // Number of goroutines with this stack trace
+	Description string
+	Labels      []*Label      `json:",omitempty"`
+	Entry       []*StackEntry `json:",omitempty"`
+}
+
+// A Process is a combined representation of a Process and a Stacktrace for the goroutines associated with it
+type Process struct {
+	PID         IDType
+	ParentPID   IDType
+	Description string
+	Start       time.Time
+	Type        string
+
+	Children []*Process `json:",omitempty"`
+	Stacks   []*Stack   `json:",omitempty"`
+}
+
+// Processes gets the processes in a thread safe manner
+func (pm *Manager) Processes(onlyRoots, noSystem bool, runInLock func()) []*Process {
+	pm.mutex.Lock()
+	processes := make([]*Process, 0, len(pm.processes))
+	if onlyRoots {
+		for _, process := range pm.processes {
+			if noSystem && process.Type == SystemProcessType {
+				continue
+			}
+			if parent, has := pm.processes[process.ParentPID]; !has ||
+				(noSystem && parent.Type == SystemProcessType) {
+				processes = append(processes, process.ToProcess(true))
+			}
+		}
+	} else {
+		for _, process := range pm.processes {
+			if noSystem && process.Type == SystemProcessType {
+				continue
+			}
+			processes = append(processes, process.ToProcess(false))
+		}
+	}
+	if runInLock != nil {
+		runInLock()
+	}
+	pm.mutex.Unlock()
+
+	sort.Slice(processes, func(i, j int) bool {
+		left, right := processes[i], processes[j]
+
+		return left.Start.Before(right.Start)
+	})
+
+	return processes
+}
+
+// ProcessStacktraces gets the processes and stacktraces in a thread safe manner
+func (pm *Manager) ProcessStacktraces(flat, onlyRequests bool) ([]*Process, int64, error) {
+	var stacks *profile.Profile
+	var err error
+	processes := pm.Processes(false, false, func() {
+		reader, writer := io.Pipe()
+		defer reader.Close()
+		go func() {
+			err := pprof.Lookup("goroutine").WriteTo(writer, 0)
+			_ = writer.CloseWithError(err)
+		}()
+		stacks, err = profile.Parse(reader)
+		if err != nil {
+			return
+		}
+	})
```

Comment:
Why does this need to be run in lock? Doesn't seem to accessing anything from `pm`

Reply:
It has to be run within pm.lock otherwise a `process` could be created between the manager becoming unlocked and the `pprof.Lookup` (and its stop the world) meaning that a goroutine could get labelled with a process that does not exist within our processes list.

If we lock the pm then any goroutines labelled with process-descriptions that are not in our process list are goroutines that are still around after the process has told us it has ended and have outlived their process. These are the goroutines that have likely escaped lifecycle control and are likely to be leaks